// @/ascii-formatting/index.js
const fg_color_codes = {
  BLACK: 30,
  RED: 31,
  GREEN: 32,
  YELLOW: 33,
  BLUE: 34,
  MAGENTA: 35,
  CYAN: 36,
  WHITE: 37,
  BRIGHT: {
    BLACK: 90,
    RED: 91,
    GREEN: 92,
    YELLOW: 93,
    BLUE: 94,
    MAGENTA: 95,
    CYAN: 96,
    WHITE: 97
  }
}

const bg_color_codes = {
  BLACK: 40,
  RED: 41,
  GREEN: 42,
  YELLOW: 43,
  BLUE: 44,
  MAGENTA: 45,
  CYAN: 46,
  WHITE: 47,
  BRIGHT: {
    BLACK: 100,
    RED: 101,
    GREEN: 102,
    YELLOW: 103,
    BLUE: 104,
    MAGENTA: 105,
    CYAN: 106,
    WHITE: 107
  }
}

/**
 * CODE OF INTEREST
 * 
 * I have tried to make this smarted but since it's an escape char, it fucks
 * things up if we try to get cute. So, just keep using the inline!
 */
const fg_colors = {
  BLACK: '\033[' + fg_color_codes.BLACK + 'm',
  RED: '\033[' + fg_color_codes.RED + 'm',
  GREEN: '\033[' + fg_color_codes.GREEN + 'm',
  YELLOW: '\033[' + fg_color_codes.YELLOW + 'm',
  BLUE: '\033[' + fg_color_codes.BLUE + 'm',
  MAGENTA: '\033[' + fg_color_codes.MAGENTA + 'm',
  CYAN: '\033[' + fg_color_codes.CYAN + 'm',
  WHITE: '\033[' + fg_color_codes.WHITE + 'm',
}

const formatting_codes = {
  UNDERLINE: 4,
  BOLD: 1,
  RESET: 0
}

const terminal_formatting = {
  UNDERLINE: '\033[' + formatting_codes.UNDERLINE + 'm',
  BOLD: '\033[' + formatting_codes.BOLD + 'm',
  RESET: '\033[' + formatting_codes.RESET + 'm',
}

module.exports = {
  colors: {
    fg: fg_colors,
    bg: bg_color_codes
  },
  formatting: terminal_formatting
}
