// @/errors/http.js
/**
 * Client Side Errors (4xx)
 */
module.exports.Client = class Client extends Error {
  code = 400
  message = 'Client Side Error'
}

module.exports.NotAuthorized = class NotAuthorized extends Error {
  code = 401
  message = 'You are not authorized to do that'
}

module.exports.Forbidden = class Forbidden extends Error {
  code = 403
  message = 'I\'m sorry Dave, I\'m afraid I can\'t let you do that'
}

module.exports.NotFound = class NotFound extends Error {
  code = 404
  message = 'Resource Not Found'
}

module.exports.MethodNotAllowed = class MethodNotAllowed extends Error {
  code = 405
  message = 'Method Not Allowed'
}

module.exports.NotAcceptable = class NotAcceptable extends Error {
  code = 406
  message = 'Request is Not Acceptable'
}

module.exports.TeaPot = class TeaPot extends Error {
  code = 418
  message = 'I am a little teapot'
}

module.exports.UnavailableForLegalReaons = class UnavailableForLegalReaons extends Error {
  code = 451
  message = 'For legal reasons outside of this server\'s control, we are not allowed to show you that resource'
}

/**
 * Server Side Errors (5xx)
 */
module.exports.Internal = class Internal extends Error {
  code = 500
  message = 'Internal Server Error'
}

module.exports.NotImplemented = class NotImplemented extends Error {
  code = 501
  message = 'The server has not implemented that yet'
}

module.exports.BadGateway = class BadGateway extends Error {
  code = 502
  message = 'Our upstream provider failed'
}
